File: src/main/java/org/modelingvalue/nelumbo/Transform.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import java.io.Serial;

import org.modelingvalue.collections.Entry;
import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.Set;
import org.modelingvalue.nelumbo.collections.NList;
import org.modelingvalue.nelumbo.patterns.Functor;
import org.modelingvalue.nelumbo.patterns.Pattern;
import org.modelingvalue.nelumbo.patterns.TokenTextPattern;
import org.modelingvalue.nelumbo.syntax.ParseContext;
import org.modelingvalue.nelumbo.syntax.ParseException;

public final class Transform extends Node {
    @Serial
    private static final long serialVersionUID = -5542746717620873208L;

    public Transform(Functor functor, List<AstElement> elements, Object... args) throws ParseException {
        super(functor, elements, args);
    }

    private Transform(Object[] array, List<AstElement> elements, Transform declaration) {
        super(array, elements, declaration);
    }

    @Override
    protected Transform struct(Object[] array, List<AstElement> elements, Node declaration) {
        return new Transform(array, elements, (Transform) declaration);
    }

    @Override
    public Transform set(int i, Object... a) {
        return (Transform) super.set(i, a);
    }

    public Node source() {
        return (Node) get(0);
    }

    @SuppressWarnings("unchecked")
    public List<Node> targets() {
        return (List<Node>) get(1);
    }

    public List<Node> targetsFlattened() {
        List<Node> result = List.of();
        for (Node e : targets()) {
            if (e instanceof NList nl) {
                result = result.addAll(nl.elements());
            } else {
                result = result.add(e);
            }
        }
        return result;
    }

    public Set<Functor> literals() {
        Set<Functor> literals = Set.of();
        for (Node target : targetsFlattened()) {
            if (target instanceof Functor functor && functor.pattern() instanceof TokenTextPattern) {
                literals = literals.add(functor);
            }
        }
        return literals;
    }

    @Override
    public Node init(KnowledgeBase knowledgeBase, ParseContext ctx) throws ParseException {
        knowledgeBase.addTransform(this);
        return this;
    }

    public Node transform(Node start, Node node, Node result, KnowledgeBase knowledgeBase, ParseContext ctx) throws ParseException {
        Map<Variable, Object> binding = node.getBinding(start);
        if (binding == null) {
            return result;
        }
        Map<Functor, Functor> functors = Map.of();
        for (Node target : targetsFlattened()) {
            if (target instanceof Functor functor && !Type.VARIABLE.isAssignableFrom(functor.resultType()) && !functor.pattern().equals(start)) {
                Functor rewrite = functor.setBinding(binding).resetDeclaration();
                for (Entry<Functor, Functor> e : functors) {
                    if (functor.equals(knowledgeBase.literal(e.getKey()))) {
                        knowledgeBase.addLiteral(e.getValue(), rewrite);
                        break;
                    }
                }
                functors = functors.put(functor, rewrite);
                rewrite.init(knowledgeBase, ctx);
                result = add(result, rewrite);
            }
        }
        if (start instanceof Pattern) {
            return result;
        }
        Map<Functor, Functor> fm = functors;
        for (Node target : targetsFlattened()) {
            if (!(target instanceof Functor)) {
                Node rewrite = target.replace(n -> {
                    if (n.typeOrFunctor() instanceof Functor f) {
                        Functor r = fm.get(f);
                        if (r != null) {
                            return n.setFunctor(r);
                        }
                    }
                    return n;
                }).setBinding(binding).setAstElements(node.astElements()).resetDeclaration();
                rewrite.init(knowledgeBase, ctx);
                result = add(result, rewrite);
            }
        }
        return result;
    }

    private static Node add(Node result, Node rewrite) {
        if (result instanceof NList list) {
            return new NList(List.of(), list, rewrite);
        } else if (result != null) {
            return new NList(Type.ROOT, List.of(result, rewrite), List.of(result, rewrite));
        }
        return null;
    }

}

File: web-ts/src/Transform.ts
/**
 * Transform - metaprogramming/rewriting.
 * Ported from Java: org.modelingvalue.nelumbo.Transform
 */

import { List, Map, Set } from 'immutable';
import type { AstElement } from './AstElement';
import { Type } from './Type';
import { Node } from './Node';
import { Pattern } from './patterns/Pattern';
import { Functor } from './patterns/Functor';
import { TokenTextPattern } from './patterns/TokenTextPattern';
import type { KnowledgeBase } from './KnowledgeBase';
import { NList } from './collections/NList';

// Re-export NList for backwards compatibility
export { NList } from './collections/NList';

/**
 * Transform - a rewrite/transformation rule.
 */
export class Transform extends Node {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, ...args);
  }

  protected static fromDataTransform(data: unknown[], declaration?: Node): Transform {
    const transform = Object.create(Transform.prototype) as Transform;
    (transform as unknown as { _data: unknown[] })._data = data;
    (transform as unknown as { _declaration: Node | undefined })._declaration = declaration ?? transform;
    return transform;
  }

  protected override struct(data: unknown[], declaration?: Node): Node {
    return Transform.fromDataTransform(data, declaration ?? this.declaration());
  }

  override set(i: number, ...a: unknown[]): Node {
    return super.set(i, ...a);
  }

  /**
   * Get the source pattern.
   */
  source(): Node {
    return this.get(0) as Node;
  }

  /**
   * Get the target nodes.
   */
  targets(): List<Node> {
    return this.get(1) as List<Node>;
  }

  /**
   * Get flattened list of targets.
   */
  targetsFlattened(): List<Node> {
    let result = List<Node>();
    for (const e of this.targets()) {
      if (e instanceof NList) {
        result = result.concat(e.elements());
      } else {
        result = result.push(e);
      }
    }
    return result;
  }

  /**
   * Get literal functors from targets.
   */
  literals(): Set<Functor> {
    let literals = Set<Functor>();
    for (const target of this.targetsFlattened()) {
      if (target instanceof Functor && target.pattern() instanceof TokenTextPattern) {
        literals = literals.add(target);
      }
    }
    return literals;
  }

  /**
   * Initialize this transform in a knowledge base.
   */
  initInKb(knowledgeBase: KnowledgeBase): Transform {
    knowledgeBase.addTransform(this);
    return this;
  }

  /**
   * Rewrite nodes matching the source pattern.
   */
  rewrite(start: Node, node: Node, knowledgeBase: KnowledgeBase): void {
    const binding = node.getBinding(start);
    if (binding === null) {
      return;
    }

    let functors = Map<Functor, Functor>();
    for (const target of this.targetsFlattened()) {
      if (target instanceof Functor && !Type.VARIABLE.isAssignableFrom(target.resultType()) && !target.pattern().equals(start)) {
        const rewrite = target.setBinding(binding).resetDeclaration() as Functor;
        for (const [key, value] of functors.entries()) {
          if (target.equals(knowledgeBase.literal(key))) {
            knowledgeBase.addLiteral(value, rewrite);
            break;
          }
        }
        functors = functors.set(target, rewrite);
        rewrite.initInKb(knowledgeBase);
      }
    }

    if (start instanceof Pattern) {
      return;
    }

    const fm = functors;
    for (const target of this.targetsFlattened()) {
      if (!(target instanceof Functor)) {
        let rewrite = target.replace(n => {
          const typeOrFunctor = n.typeOrFunctor();
          if (typeOrFunctor instanceof Functor) {
            const r = fm.get(typeOrFunctor);
            if (r !== undefined) {
              return n.setFunctor(r);
            }
          }
          return n;
        });
        rewrite = rewrite.setBinding(binding).setAstElements(node.astElements()).resetDeclaration();
        if (rewrite instanceof Functor) {
          rewrite.initInKb(knowledgeBase);
        }
      }
    }
  }
}
